isis/src/base/objs/SpicePosition/SpicePosition.cpp

```diff
@@ -359,25 +368,28 @@ namespace Isis {
     p_fullCacheSize = isdPos["spk_table_original_size"].get<double>();
     p_cacheTime = isdPos["ephemeris_times"].get<std::vector<double>>();
 
-    for (auto it = isdPos["positions"].begin(); it != isdPos["positions"].end(); it++) {
-      std::vector<double> pos = {it->at(0).get<double>(), it->at(1).get<double>(), it->at(2).get<double>()};
-      p_cache.push_back(pos);
-    }
-
-    p_cacheVelocity.clear();
-
     bool hasVelocityKey = isdPos.find("velocities") != isdPos.end();
 
+    std::vector<ale::State> stateCache;
     if (hasVelocityKey) {
-      for (auto it = isdPos["velocities"].begin(); it != isdPos["velocities"].end(); it++) {
-        std::vector<double> vel = {it->at(0).get<double>(), it->at(1).get<double>(), it->at(2).get<double>()};
-        p_cacheVelocity.push_back(vel);
+      for (auto it = isdPos["positions"].begin(); it != isdPos["positions"].end(); it++) {
+        std::vector<double> pos = {it->at(0).get<double>(), it->at(1).get<double>(), it->at(2).get<double>()};
+        int index = it - isdPos["positions"].begin();
+        std::vector<double> vel = isdPos["velocities"][index];
+        stateCache.push_back(ale::State(pos, vel));
       }
     }
+    else {
+      for (auto it = isdPos["positions"].begin(); it != isdPos["positions"].end(); it++) {
+        std::vector<double> pos = {it->at(0).get<double>(), it->at(1).get<double>(), it->at(2).get<double>()};
```

Comment:
These positions are still being setup different from the velocities.

```
std::vector<double> vel = isdPos["velocities"][index];
```

vs

```
std::vector<double> pos = {it->at(0).get<double>(), it->at(1).get<double>(), it->at(2).get<double>()};
```

Reply:
But it's iterating over positions, right? How would your second example work without creating a second iterator?

Reply:
I am mostly concerned about using at().get for one and straight indexing for the other. I would rather the position use `it->get<std::vector<double>>()`.

Reply:
Thanks @jessemapel, I had been focusing on the wrong part of your comment!